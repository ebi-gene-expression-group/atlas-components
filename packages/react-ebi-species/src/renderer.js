"use strict";

//*------------------------------------------------------------------*

var React = require('react');
var ReactDOM = require('react-dom');

//*------------------------------------------------------------------*

var Icon = require('./SpeciesIcon.jsx');

//*------------------------------------------------------------------*

module.exports = function(mountNode,species,colourOverride, colourPerKingdom) {
    ReactDOM.render(
        React.createElement(Icon,{
          species: species,
          colourOverride: colourOverride,
          colourPerKingdom: colourPerKingdom
        }), mountNode
    );
};
