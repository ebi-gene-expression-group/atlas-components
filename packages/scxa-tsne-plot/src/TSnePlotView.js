import React from 'react'
import PropTypes from 'prop-types'

import URI from 'urijs'

import ClusterTSnePlot from './ClusterTSnePlot'
import GeneExpressionTSnePlot from './GeneExpressionTSnePlot'

const fetchResponseJson = async (base, endpoint) => {
  const response = await fetch(URI(endpoint, base).toString())
  const responseJson = await response.json()
  return responseJson
}

class TSnePlotView extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      geneExpressionData: {
        max: null,
        min: null,
        series: [],
        unit: ``
      },
      cellClustersData: {
        series: []
      },
      geneExpressionErrorMessage: null,
      cellClustersErrorMessage: null,
      loadingCellClusters: false,
      loadingGeneExpression: false
    }
  }

  _fetchAndSetStateCellClusters({atlasUrl, experimentAccession, selectedColourBy, selectedColourByCategory, selectedPerplexity}) {
    this.setState({
      loadingCellClusters: true
    }, () => {
      let endpoint
      if(selectedColourByCategory === `clusters`) {
        endpoint = `json/experiments/${experimentAccession}/tsneplot/${selectedPerplexity}/clusters/k/${selectedColourBy}`
      }
      else if(selectedColourByCategory === `metadata`) {
        endpoint = `json/experiments/${experimentAccession}/tsneplot/${selectedPerplexity}/metadata/${selectedColourBy}`
      }
      fetchResponseJson(atlasUrl, endpoint)
        .then((responseJson) => {
          this.setState({
            cellClustersData: responseJson,
            cellClustersErrorMessage: null,
            loadingCellClusters: false
          })
        })
        .catch((reason) => {
          this.setState({
            cellClustersErrorMessage: `${reason.name}: ${reason.message}`,
            loadingCellClusters: false
          })
        })
    })
  }

  _fetchAndSetStateGeneId({atlasUrl, experimentAccession, selectedPerplexity, geneId}) {
    this.setState({
      loadingGeneExpression: true
    }, () => {
      fetchResponseJson(atlasUrl, `json/experiments/${experimentAccession}/tsneplot/${selectedPerplexity}/expression/${geneId}`)
        .then((responseJson) => {
          this.setState({
            geneExpressionData: responseJson,
            geneExpressionErrorMessage: null,
            loadingGeneExpression: false,
          })
        })
        .catch((reason) => {
          this.setState({
            geneExpressionErrorMessage: `${reason.name}: ${reason.message}`,
            loadingGeneExpression: false
          })
        })
    })
  }

  componentDidUpdate(previousProps) {
    if (previousProps.selectedPerplexity !== this.props.selectedPerplexity || previousProps.experimentAccession !== this.props.experimentAccession) {
      this._fetchAndSetStateCellClusters(this.props)
      this._fetchAndSetStateGeneId(this.props)
    } else if (previousProps.selectedColourByCategory !== this.props.selectedColourBy && previousProps.selectedColourBy !== this.props.selectedColourBy) {
      this._fetchAndSetStateCellClusters(this.props)
    } else if (previousProps.geneId !== this.props.geneId) {
      this._fetchAndSetStateGeneId(this.props)
    }
  }

  componentDidMount() {
    this._fetchAndSetStateCellClusters(this.props)
    this._fetchAndSetStateGeneId(this.props)
  }

  render() {
    const {height, atlasUrl, resourcesUrl, suggesterEndpoint} = this.props
    const {wrapperClassName, clusterPlotClassName, expressionPlotClassName} = this.props
    const {geneId, speciesName, highlightClusters} = this.props
    const {ks, perplexities, selectedPerplexity, metadata, selectedColourBy} = this.props
    const {onChangePerplexity, onSelectGeneId, onChangeColourBy} = this.props
    const {loadingGeneExpression, geneExpressionData, geneExpressionErrorMessage} = this.state
    const {loadingCellClusters, cellClustersData, cellClustersErrorMessage} = this.state

    const getTooltipContent = (cellId) => {
      return fetchResponseJson(atlasUrl, `json/experiment/${this.props.experimentAccession}/cell/${cellId}/metadata`)
    }

    return (
      <div className={wrapperClassName}>
        <div className={clusterPlotClassName}>
          <ClusterTSnePlot
            height={height}
            plotData={cellClustersData}
            perplexities={perplexities}
            selectedPerplexity={selectedPerplexity}
            onChangePerplexity={onChangePerplexity}
            ks={ks}
            metadata={metadata}
            onChangeColourBy={onChangeColourBy}
            selectedColourBy={selectedColourBy}
            highlightClusters={highlightClusters}
            loading={loadingCellClusters}
            resourcesUrl={resourcesUrl}
            errorMessage={cellClustersErrorMessage}
            tooltipContent={getTooltipContent}
          />
        </div>

        <div className={expressionPlotClassName}>
          <GeneExpressionTSnePlot
            height={height}
            plotData={geneExpressionData}
            atlasUrl={atlasUrl}
            suggesterEndpoint={suggesterEndpoint}
            onSelectGeneId={onSelectGeneId}
            geneId={geneId}
            speciesName={speciesName}
            highlightClusters={[]}
            loading={loadingGeneExpression}
            resourcesUrl={resourcesUrl}
            errorMessage={geneExpressionErrorMessage}
          />
        </div>
      </div>
    )
  }

  componentDidCatch(error) {
    this.setState({
      errorMessage: `${error}`
    })
  }
}

TSnePlotView.propTypes = {
  atlasUrl: PropTypes.string.isRequired,
  wrapperClassName: PropTypes.string,
  clusterPlotClassName: PropTypes.string,
  expressionPlotClassName: PropTypes.string,
  suggesterEndpoint: PropTypes.string.isRequired,
  experimentAccession: PropTypes.string.isRequired,
  ks: PropTypes.arrayOf(PropTypes.number).isRequired,
  perplexities: PropTypes.arrayOf(PropTypes.number).isRequired,
  selectedPerplexity: PropTypes.number.isRequired,

  metadata: PropTypes.arrayOf(PropTypes.shape({
    value: PropTypes.string,
    label: PropTypes.string
  })),
  selectedColourBy: PropTypes.string,
  selectedColourByCategory: PropTypes.string,
  onChangeColourBy: PropTypes.func,

  highlightClusters: PropTypes.arrayOf(PropTypes.number),
  geneId: PropTypes.string.isRequired,
  speciesName: PropTypes.string.isRequired,
  height: PropTypes.number,
  resourcesUrl: PropTypes.string,
  onSelectGeneId: PropTypes.func,
  onChangePerplexity: PropTypes.func
}

TSnePlotView.defaultProps = {
  highlightClusters: [],
  wrapperClassName: `row`,
  clusterPlotClassName: `small-12 medium-6 columns`,
  expressionPlotClassName: `small-12 medium-6 columns`,
  geneId: ``,
  speciesName: ``,
  height: 800,
  onSelectGeneId: () => {},
  onChangeColourBy: () => {},
  onPerplexityChange: () => {}
}

export default TSnePlotView
