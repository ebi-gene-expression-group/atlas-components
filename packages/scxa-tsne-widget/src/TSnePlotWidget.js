import React from 'react'
import PropTypes from 'prop-types'

import URI from 'urijs'

import TSnePlotView from '@ebi-gene-expression-group/scxa-tsne-plot'

class TSnePlotWidget extends React.Component {
  constructor(props) {
    super(props)
    const defaultPlotMethodAndParameterisation = this.props.defaultPlotMethodAndParameterisation
    this.state = {
      selectedPlotType: Object.keys(defaultPlotMethodAndParameterisation)[0],
      selectedPlotOption: Object.values(Object.values(defaultPlotMethodAndParameterisation)[0])[0],
      selectedPlotOptionLabel: Object.keys(Object.values(defaultPlotMethodAndParameterisation)[0])[0] + ": "
          + Object.values(Object.values(defaultPlotMethodAndParameterisation)[0])[0],
      selectedColourBy: this.props.metadata ? this.props.metadata[0].value : ``,
      selectedGeneId: this.props.geneId,
      selectedColourByCategory: `metadata`,
      metadataErrorMessage: null,
      loadingMetadata: false,
      geneId: ``,
      highlightClusters: []
    }

    this.experimentAccessionInput = React.createRef()
    this.highlightClustersInput = React.createRef()

    this._handleSubmit = this._handleSubmit.bind(this)
  }
  _resetHighlightClusters() {
    this.setState({
      highlightClusters: []
    })
    if (this.highlightClustersInput.current) {
      this.highlightClustersInput.current.value = ``
    }
  }

  _handleSubmit(event) {
    const highlightClustersValue = this.highlightClustersInput.current ? this.highlightClustersInput.current.value : ''
    const highlightClusters = highlightClustersValue.split(',')
        .map((e) => parseInt(e.trim()))
        .filter((e) => !isNaN(e))

    this.setState({
      experimentAccession: this.experimentAccessionInput.current ? this.experimentAccessionInput.current.value : '',
      highlightClusters: highlightClusters
    })

    event.preventDefault()
  }

  render() {
    const {height, atlasUrl, suggesterEndpoint} = this.props
    const {wrapperClassName, clusterPlotClassName, expressionPlotClassName} = this.props
    const {geneIds, speciesName, ks, experimentAccession, showControls} = this.props
    const {plotTypesAndOptions, defaultPlotMethodAndParameterisation, metadata} = this.props
    const {selectedColourBy, selectedColourByCategory, loadingMetadata, selectedGeneId} = this.state

    const plotTypeDropdown =  [
      {
        plotType: Object.keys(defaultPlotMethodAndParameterisation)[0],
        plotOptions: plotTypesAndOptions[Object.keys(defaultPlotMethodAndParameterisation)[0]]
      },
      {
        plotType: Object.keys(defaultPlotMethodAndParameterisation)[1],
        plotOptions: plotTypesAndOptions[Object.keys(defaultPlotMethodAndParameterisation)[1]]
      }
    ]

    return (
      loadingMetadata ?
        <p className={`row column loading-message`}>Loading, please waitâ€¦</p> :
        <React.Fragment>
          <TSnePlotView
            atlasUrl={atlasUrl}
            suggesterEndpoint={suggesterEndpoint}
            wrapperClassName={wrapperClassName}
            clusterPlotClassName={clusterPlotClassName}
            expressionPlotClassName={expressionPlotClassName}
            speciesName={speciesName}
            experimentAccession={experimentAccession}
            ks={ks}
            metadata={metadata}
            plotTypeDropdown={plotTypeDropdown}
            selectedColourBy={selectedColourBy}
            selectedColourByCategory={selectedColourByCategory}
            onChangePerplexity={() => {}}
            geneId={selectedGeneId}
            geneIds={geneIds}
            height={height}
            showControls={showControls}
            selectedPlotOption={this.state.selectedPlotOption}
            selectedPlotType={this.state.selectedPlotType}
            selectedPlotOptionLabel={this.state.selectedPlotOptionLabel}
            onChangePlotTypes={(plotOption) => {
              this.setState({
                selectedPlotType: plotOption.value,
                selectedPlotOption: Object.values(defaultPlotMethodAndParameterisation[plotOption.value])[0],
                selectedPlotOptionLabel: Object.keys(defaultPlotMethodAndParameterisation[plotOption.value])[0]
                    + ": " + Object.values(defaultPlotMethodAndParameterisation[plotOption.value])[0],
              })
            }}
            onChangePlotOptions={(plotOption) => {
              this.setState({
                selectedPlotOption: plotOption.value,
                selectedPlotOptionLabel: plotOption.label
              })
            }}
            highlightClusters={this.state.highlightClusters}
            onChangeColourBy={(colourByCategory, colourByValue) => {
              this.setState({
                selectedColourBy : colourByValue,
                selectedColourByCategory : colourByCategory
              })
              this._resetHighlightClusters()
            }}
            onSelectGeneId={(geneId) => {
              this.setState({ geneId: geneId })
              this._resetHighlightClusters()
            }}
          />
          <p>
            To know more about this experiment please go to <a target={`_blank`} href={`https://www.ebi.ac.uk/gxa/sc/experiments/${experimentAccession}`}>Single Cell Expression Atlas </a>.
          </p>
        </React.Fragment>
    )
  }
}

TSnePlotWidget.propTypes = {
  atlasUrl: PropTypes.string,
  wrapperClassName: PropTypes.string,
  clusterPlotClassName: PropTypes.string,
  expressionPlotClassName: PropTypes.string,
  experimentAccession: PropTypes.string.isRequired,
  geneId: PropTypes.string.isRequired,
  geneIds: PropTypes.arrayOf(PropTypes.string),
  ks: PropTypes.arrayOf(PropTypes.number),
  suggesterEndpoint: PropTypes.string,
  speciesName: PropTypes.string,
  height: PropTypes.number,
  showControls: PropTypes.bool
}


TSnePlotWidget.defaultProps = {
  atlasUrl: `https://www.ebi.ac.uk/gxa/sc/`,
  suggesterEndpoint: `json/suggestions`,
  wrapperClassName: `row expanded`,
  clusterPlotClassName: `small-12 large-6 columns`,
  expressionPlotClassName: `small-12 large-6 columns`,
  speciesName: ``,
  height: 800,
  ks: [],
  geneIds: [],
  showControls: false
}

export default TSnePlotWidget
