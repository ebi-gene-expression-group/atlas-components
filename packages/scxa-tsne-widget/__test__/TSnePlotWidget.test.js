import '@babel/polyfill'
import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'

import TSnePlotWidget from '../src/TSnePlotWidget'
import TSnePlotView from '@ebi-gene-expression-group/scxa-tsne-plot'

describe(`TSnePlotWidget`, () => {

  test(`whether showControl prop is true or false`, () => {
    const experimentAccession = `E-MTAB-5061`
    const speciesName = `Homo Sapiens`

    const onChangeColourBy = () => {}
    const onChangePerplexity = () => {}
    const plotData = {
      series: []
    }

    const wrapper =
      shallow(
        <TSnePlotWidget
          speciesName={speciesName}
          experimentAccession={experimentAccession}
          ks={[]}
          metadata={[]}
          selectedColourBy={`0`}
          onChangeColourBy={onChangeColourBy}
          perplexities={[]}
          selectedPerplexity={0}
          onChangePerplexity={onChangePerplexity}
          loading={true}
          plotData={plotData}/>)

    const tSnePlotView = wrapper.find(TSnePlotView)

    expect(tSnePlotView.props().showControls).toBe(false)
  })

  test(`whether geneIds prop is passed or not`, () => {
    const experimentAccession = `E-MTAB-5061`
    const speciesName = `Homo Sapiens`

    const onChangeColourBy = () => {}
    const onChangePerplexity = () => {}
    const geneIds = [`foo`, `bar`]
    const plotData = {
      series: []
    }

    const wrapper =
        shallow(
            <TSnePlotWidget
                speciesName={speciesName}
                experimentAccession={experimentAccession}
                ks={[]}
                metadata={[]}
                geneIds={geneIds}
                selectedColourBy={`0`}
                onChangeColourBy={onChangeColourBy}
                perplexities={[]}
                selectedPerplexity={0}
                onChangePerplexity={onChangePerplexity}
                loading={true}
                plotData={plotData}/>)

    const tSnePlotView = wrapper.find(TSnePlotView)

    expect(tSnePlotView.props().geneIds.length).toEqual(2)
  })

  /*test(`matches snapshot`, () => {
    const experimentAccession = `E-MTAB-5061`
    const speciesName = `Homo Sapiens`

    const onChangeColourBy = () => {}
    const onChangePerplexity = () => {}
    const plotData = {
      series: []
    }

    const tree = renderer
      .create(
        <TSnePlotWidget
          speciesName={speciesName}
          experimentAccession={experimentAccession}
          ks={[]}
          geneId={``}
          metadata={[]}
          selectedColourBy={`0`}
          onChangeColourBy={onChangeColourBy}
          perplexities={[]}
          selectedPerplexity={0}
          onChangePerplexity={onChangePerplexity}
          loading={true}
          plotData={plotData}/>)
      .toJSON()

    expect(tree).toMatchSnapshot()
  })*/
})
