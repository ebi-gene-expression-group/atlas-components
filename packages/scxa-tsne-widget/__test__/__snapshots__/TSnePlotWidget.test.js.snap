// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TSnePlotWidget matches snapshot 1`] = `
"<Fragment>
  <TSnePlotView atlasUrl=\\"https://www.ebi.ac.uk/gxa/sc/\\" suggesterEndpoint=\\"json/suggestions\\" wrapperClassName=\\"row expanded\\" clusterPlotClassName=\\"small-12 large-6 columns\\" expressionPlotClassName=\\"small-12 large-6 columns\\" speciesName=\\"Homo Sapiens\\" experimentAccession=\\"E-MTAB-5061\\" ks={{...}} metadata={{...}} selectedColourBy=\\"inferred_cell_type\\" selectedColourByCategory=\\"metadata\\" perplexities={{...}} selectedPerplexity={[undefined]} onChangePerplexity={[Function: onChangePerplexity]} geneId=\\"\\" geneIds={{...}} height={800} onChangeColourBy={[Function: onChangeColourBy]} onSelectGeneId={[Function: onSelectGeneId]} showControls={false} highlightClusters={{...}} onPerplexityChange={[Function: onPerplexityChange]} />
  <p>
    To know more about this experiment please go to 
    <a target=\\"_blank\\" href=\\"https://www.ebi.ac.uk/gxa/sc/experiments/E-MTAB-5061\\">
      Single Cell Expression Atlas 
    </a>
    .
  </p>
</Fragment>"
`;
